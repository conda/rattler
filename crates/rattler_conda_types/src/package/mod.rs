//! Contains models of files that are found in the `info/` directory of a package.

mod about;
mod archive_identifier;
mod archive_type;
mod entry_point;
mod files;
mod has_prefix;
mod index;
mod link;
mod no_link;
mod no_softlink;
mod package_metadata;
mod paths;
mod run_exports;

use std::io::Read;
use std::path::Path;
pub use {
    about::AboutJson,
    archive_identifier::ArchiveIdentifier,
    archive_type::ArchiveType,
    entry_point::EntryPoint,
    files::Files,
    has_prefix::HasPrefix,
    has_prefix::HasPrefixEntry,
    index::IndexJson,
    link::{LinkJson, NoArchLinks, PythonEntryPoints},
    no_link::NoLink,
    no_softlink::NoSoftlink,
    package_metadata::PackageMetadata,
    paths::{FileMode, PathType, PathsEntry, PathsJson, PrefixPlaceholder},
    run_exports::RunExportsJson,
};

/// A trait implemented for structs that represent specific files in a Conda archive.
///
/// This trait provides a standardised interface for accessing the contents of known files in a
/// Conda package, such as the `index.json` (see [`IndexJson`]) or `about.json` (see [`AboutJson`])
/// files. Structs that represent these files should implement this trait in order to ensure that
/// they can be easily accessed and manipulated by other code that expects a consistent interface.
pub trait PackageFile: Sized {
    /// Returns the path to the file within the Conda archive.
    ///
    /// The path is relative to the root of the archive and include any necessary directories.
    fn package_path() -> &'static Path;

    /// Parses the object from a string, using a format appropriate for the file type.
    ///
    /// For example, if the file is in JSON format, this function parses the JSON string and returns
    /// the resulting object. If the file is not in a parsable format, this function returns an
    /// error.
    fn from_str(str: &str) -> Result<Self, std::io::Error>;

    /// Parses the object from a `Read` trait object, using a format appropriate for the file type.
    ///
    /// For example, if the file is in JSON format, this function reads the data from the `Read`
    /// object, parse the JSON string and return the resulting object. If the file is not in a
    /// parsable format, this function returns an error.
    fn from_reader(mut reader: impl Read) -> Result<Self, std::io::Error> {
        let mut str = String::new();
        reader.read_to_string(&mut str)?;
        Self::from_str(&str)
    }

    /// Parses the object from a file specified by a `path`, using a format appropriate for the file
    /// type.
    ///
    /// For example, if the file is in JSON format, this function reads the data from the file at
    /// the specified path, parse the JSON string and return the resulting object. If the file is
    /// not in a parsable format or if the file could not read, this function returns an error.
    fn from_path(path: impl AsRef<Path>) -> Result<Self, std::io::Error> {
        Self::from_str(&fs_err::read_to_string(path)?)
    }

    /// Parses the object by looking up the appropriate file from the root of the specified Conda
    /// archive directory, using a format appropriate for the file type.
    ///
    /// For example, if the file is in JSON format, this function reads the appropriate file from
    /// the archive, parse the JSON string and return the resulting object. If the file is not in a
    /// parsable format or if the file could not be read, this function returns an error.
    fn from_package_directory(path: impl AsRef<Path>) -> Result<Self, std::io::Error> {
        Self::from_path(path.as_ref().join(Self::package_path()))
    }
}
