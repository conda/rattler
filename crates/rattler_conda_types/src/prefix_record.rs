//! Defines the `[PrefixRecord]` struct.

use crate::package::FileMode;
use crate::repo_data::RecordFromPath;
use crate::repo_data_record::RepoDataRecord;
use crate::{menuinst, PackageRecord};
use rattler_digest::serde::SerializableHash;
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};
use serde_with::serde_as;
use std::io::{BufWriter, Read};
use std::path::{Path, PathBuf};
use std::str::FromStr;
use tempfile::NamedTempFile;

#[cfg(feature = "rayon")]
use rayon::prelude::*;

/// Information about every file installed with the package.
///
/// This struct is similar to the [`crate::package::PathsJson`] struct. The difference is that this
/// information refers to installed files whereas [`crate::package::PathsJson`] describes the
/// instructions on how to install a package.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct PrefixPaths {
    /// The version of the file
    pub paths_version: u64,

    /// All entries included in the package.
    #[serde(default)]
    pub paths: Vec<PathsEntry>,
}

impl Default for PrefixPaths {
    fn default() -> Self {
        Self {
            paths_version: 1,
            paths: Vec::default(),
        }
    }
}

impl From<Vec<PathsEntry>> for PrefixPaths {
    fn from(paths: Vec<PathsEntry>) -> Self {
        Self {
            paths,
            ..Default::default()
        }
    }
}

/// Information about a single file installed for a package.
///
/// This struct is similar to the [`crate::package::PathsEntry`] struct. The difference is that this
/// information refers to installed files whereas [`crate::package::PathsEntry`] describes the
/// instructions on how to install a file.
#[serde_as]
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Hash)]
pub struct PathsEntry {
    /// The relative path from the root of the package
    #[serde(rename = "_path")]
    #[serde_as(as = "crate::utils::serde::NormalizedPath")]
    pub relative_path: PathBuf,

    /// The original path of the file in the package. This is only set if the file was clobbered by
    /// another package and therefore the original path is not the same as the relative path.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub original_path: Option<PathBuf>,

    /// How the file was installed into the environment
    pub path_type: PathType,

    /// Whether or not this file should not be linked
    #[serde(
        default = "no_link_default",
        skip_serializing_if = "is_no_link_default"
    )]
    pub no_link: bool,

    /// A hex representation of the SHA256 hash of the contents of the file.
    #[serde_as(as = "Option<SerializableHash::<rattler_digest::Sha256>>")]
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sha256: Option<rattler_digest::Sha256Hash>,

    /// A hex representation of the SHA256 hash of the original file from which this was created.
    #[serde_as(as = "Option<SerializableHash::<rattler_digest::Sha256>>")]
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sha256_in_prefix: Option<rattler_digest::Sha256Hash>,

    /// The size of the file in bytes
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub size_in_bytes: Option<u64>,

    /// The file mode of the entry. This is used in conjunction with a prefix_placeholder
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub file_mode: Option<FileMode>,

    /// The original sentinel value used for prefix-replacement from the package
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub prefix_placeholder: Option<String>,
}

/// Information about a single file installed for a package.
///
/// This enum is similar to the [`crate::package::PathType`] enum. This enum includes more enum entries
/// that are created when a file is installed.
#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Hash)]
#[serde(rename_all = "snake_case")]
pub enum PathType {
    /// The file was installed as a hard link (i.e. a file that points to another file in the package cache)
    #[serde(rename = "hardlink")]
    HardLink,
    #[serde(rename = "softlink")]
    /// The file was installed as a soft link (i.e. a soft link to a file in the package cache)
    SoftLink,
    /// An empty directory was created at installation time here
    Directory,
    /// This is a file that was automatically "compiled" from Python source code when a `noarch` package
    /// was installed
    PycFile,
    /// This file is a Python entry point script for Windows (a `<entrypoint>-script.py` Python script file)
    WindowsPythonEntryPointScript,
    /// This file is a Python entry point executable for Windows (a `<entrypoint>.exe` file)
    WindowsPythonEntryPointExe,
    /// This file is a Python entry point executable for Unix (a `<entrypoint>` Python script file)
    /// Entry points are created in the `bin/...` directory when installing Python noarch packages
    UnixPythonEntryPoint,
    /// NOT USED - path to the package's .json file in conda-meta
    LinkedPackageRecord,
}

impl From<crate::package::PathType> for PathType {
    fn from(ty: crate::package::PathType) -> Self {
        match ty {
            crate::package::PathType::HardLink => PathType::HardLink,
            crate::package::PathType::SoftLink => PathType::SoftLink,
            crate::package::PathType::Directory => PathType::Directory,
        }
    }
}

impl RecordFromPath for PrefixRecord {
    fn from_path(path: &Path) -> Result<Self, std::io::Error> {
        Self::from_path(path)
    }
}

/// A record of a single package installed within an environment. The struct includes the
/// [`RepoDataRecord`] which specifies information about where the original package comes from.
#[serde_as]
#[derive(Debug, Deserialize, Serialize, Eq, PartialEq, Clone)]
pub struct PrefixRecord {
    /// The data from the repodata
    #[serde(flatten)]
    pub repodata_record: RepoDataRecord,

    /// The path to where the archive of the package was stored on disk.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub package_tarball_full_path: Option<PathBuf>,

    /// The path that contains the extracted package content.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub extracted_package_dir: Option<PathBuf>,

    /// A sorted list of all files included in this package
    #[serde(default)]
    #[serde_as(as = "Vec<crate::utils::serde::NormalizedPath>")]
    pub files: Vec<PathBuf>,

    /// Information about how files have been linked when installing the package.
    #[serde(default)]
    pub paths_data: PrefixPaths,

    /// This field contains a reference to the package cache from where the package was linked.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub link: Option<Link>,

    /// The spec that was used when this package was installed. Note that this field is not updated if the
    /// currently another spec was used. Note: conda seems to serialize a "None" string value instead of `null`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub requested_spec: Option<String>,

    /// If menuinst is enabled and added menu items, this field contains the menuinst tracker data.
    /// This data is used to remove the menu items when the package is uninstalled.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub installed_system_menus: Vec<menuinst::Tracker>,
}

impl PrefixRecord {
    /// Parses a `paths.json` file from a reader.
    pub fn from_reader(mut reader: impl Read) -> Result<Self, std::io::Error> {
        let mut str = String::new();
        reader.read_to_string(&mut str)?;
        simd_json::serde::from_slice(&mut str.into_bytes()).map_err(Into::into)
    }

    /// Creates a `PrefixRecord` from a `RepoDataRecord`.
    pub fn from_repodata_record(
        repodata_record: RepoDataRecord,
        package_tarball_full_path: Option<PathBuf>,
        extracted_package_dir: Option<PathBuf>,
        paths: Vec<PathsEntry>,
        requested_spec: Option<String>,
        link: Option<Link>,
    ) -> Self {
        Self {
            repodata_record,
            package_tarball_full_path,
            extracted_package_dir,
            files: paths
                .iter()
                .map(|entry| entry.relative_path.clone())
                .collect(),
            paths_data: paths.into(),
            link,
            requested_spec,
            installed_system_menus: Vec::new(),
        }
    }

    /// Parses a `paths.json` file from a file.
    pub fn from_path(path: impl AsRef<Path>) -> Result<Self, std::io::Error> {
        Self::from_str(&fs_err::read_to_string(path.as_ref())?)
    }

    /// Return the canonical file name for a `PrefixRecord`. Takes the form of
    /// `<package_name>-<version>-<build>.json`.
    pub fn file_name(&self) -> String {
        format!(
            "{}-{}-{}.json",
            self.repodata_record.package_record.name.as_normalized(),
            self.repodata_record.package_record.version,
            self.repodata_record.package_record.build
        )
    }

    /// Writes the contents of this instance to the file at the specified location.
    pub fn write_to_path(
        &self,
        path: impl AsRef<Path>,
        pretty: bool,
    ) -> Result<(), std::io::Error> {
        let path = path.as_ref();
        let parent = path.parent().ok_or_else(|| {
            std::io::Error::new(
                std::io::ErrorKind::Other,
                format!(
                    "Failed to get parent directory of path '{}'",
                    path.display()
                ),
            )
        })?;

        // Use a temporary file in the same directory for atomic writes
        let temp_file = NamedTempFile::with_prefix_in("prefix_record_", parent).map_err(|e| {
            std::io::Error::new(
                std::io::ErrorKind::Other,
                format!(
                    "Failed to create temporary file in '{}': {}",
                    parent.display(),
                    e
                ),
            )
        })?;

        // Write to temp file with buffered writer
        let writer = BufWriter::with_capacity(64 * 1024, &temp_file);
        self.write_to(writer, pretty)?;

        // Make sure that all data is written to disk
        temp_file.as_file().sync_all()?;

        // Atomically rename the temp file to the target path
        temp_file.persist(path).map_err(|e| {
            std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Failed to persist file {}: {}", path.display(), e),
            )
        })?;

        Ok(())
    }

    /// Writes the contents of this instance to the file at the specified location.
    pub fn write_to(
        &self,
        writer: impl std::io::Write,
        pretty: bool,
    ) -> Result<(), std::io::Error> {
        if pretty {
            serde_json::to_writer_pretty(writer, self)?;
        } else {
            serde_json::to_writer(writer, self)?;
        }
        Ok(())
    }

    /// Collects all `PrefixRecord`s from the specified prefix. This function will read all files in
    /// the `$PREFIX/conda-meta` directory and parse them as `PrefixRecord`s.
    pub fn collect_from_prefix<T: RecordFromPath + Send>(
        prefix: &Path,
    ) -> Result<Vec<T>, std::io::Error> {
        let conda_meta_path = prefix.join("conda-meta");

        if !conda_meta_path.exists() {
            return Ok(Vec::new());
        }

        // Collect paths first to avoid holding the directory iterator during parallel processing
        let json_paths: Vec<_> = fs_err::read_dir(&conda_meta_path)?
            .filter_map(|entry| {
                entry.ok().and_then(|e| {
                    if e.file_type().ok()?.is_file()
                        && e.file_name().to_string_lossy().ends_with(".json")
                    {
                        Some(e.path())
                    } else {
                        None
                    }
                })
            })
            .collect();

        #[cfg(feature = "rayon")]
        {
            // Process files in parallel
            json_paths
                .par_iter()
                .map(|path| RecordFromPath::from_path(path))
                .collect()
        }

        #[cfg(not(feature = "rayon"))]
        {
            json_paths
                .iter()
                .map(|path| RecordFromPath::from_path(path))
                .collect()
        }
    }
}

impl FromStr for PrefixRecord {
    type Err = std::io::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        serde_json::from_str(s).map_err(Into::into)
    }
}

/// A record of a single file that was installed into the prefix
#[derive(Debug, Deserialize, Serialize, Eq, PartialEq, Clone)]
pub struct Link {
    /// The path to the file source that was installed
    pub source: PathBuf,

    /// The link type that was used to install the file
    #[serde(rename = "type")]
    pub link_type: Option<LinkType>,
}

/// The different link types that are used
#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize_repr, Deserialize_repr, Hash)]
#[serde(rename_all = "lowercase")]
#[repr(u8)]
pub enum LinkType {
    /// Hard link refers to the same inode as the source file
    HardLink = 1,
    /// Soft link is a soft link to the source file
    SoftLink = 2,
    /// Copy is a proper copy of the source file (duplicated data on hard disk)
    Copy = 3,
    /// Directory is a (empty) directory
    Directory = 4,
}

/// Returns the default value for the `no_link` value of a [`PathsEntry`]
fn no_link_default() -> bool {
    false
}

/// Returns true if the value is equal to the default value for the `no_link` value of a [`PathsEntry`]
fn is_no_link_default(value: &bool) -> bool {
    *value == no_link_default()
}

impl AsRef<RepoDataRecord> for PrefixRecord {
    fn as_ref(&self) -> &RepoDataRecord {
        &self.repodata_record
    }
}

impl AsRef<PackageRecord> for PrefixRecord {
    fn as_ref(&self) -> &PackageRecord {
        &self.repodata_record.package_record
    }
}

#[cfg(test)]
mod test {
    use crate::get_test_data_dir;
    use rstest::rstest;

    #[rstest]
    #[case::xz_5_2_6_h8d14728_0("xz-5.2.6-h8d14728_0.json")]
    #[case::libsqlite_3_40_0_hcfcfb64_0("libsqlite-3.40.0-hcfcfb64_0.json")]
    #[case::menuinst_1_4_19_py311h1ea47a8_1("menuinst-1.4.19-py311h1ea47a8_1.json")]
    #[case::pip_23_0_pyhd8ed1ab_0_json("pip-23.0-pyhd8ed1ab_0.json")]
    #[case::pysocks_1_7_1_pyh0701188_6("pysocks-1.7.1-pyh0701188_6.json")]
    #[case::requests_2_28_2_pyhd8ed1ab_0("requests-2.28.2-pyhd8ed1ab_0.json")]
    #[case::tk_8_6_12_h8ffe710_0("tk-8.6.12-h8ffe710_0.json")]
    #[case::urllib3_1_26_14_pyhd8ed1ab_0("urllib3-1.26.14-pyhd8ed1ab_0.json")]
    #[case::vc_14_3_hb6edc58_10_json("vc-14.3-hb6edc58_10.json")]
    #[case::wheel_0_38_4_pyhd8ed1ab_0("wheel-0.38.4-pyhd8ed1ab_0.json")]
    fn parse_prefix_record(#[case] path_name: &str) {
        let path = get_test_data_dir().join("conda-meta").join(path_name);
        let prefix_record = super::PrefixRecord::from_path(path).unwrap();
        insta::assert_yaml_snapshot!(path_name.replace('.', "_"), prefix_record);
    }
}
