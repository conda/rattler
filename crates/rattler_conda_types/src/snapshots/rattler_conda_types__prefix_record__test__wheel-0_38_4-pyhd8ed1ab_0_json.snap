---
source: crates/rattler_conda_types/src/prefix_record.rs
expression: prefix_record
---
build: pyhd8ed1ab_0
build_number: 0
depends:
  - python >=3.7
license: MIT
license_family: MIT
md5: c829cfb8cb826acb9de0ac1a2df0a940
name: wheel
noarch: python
sha256: bd4f11ff075ff251ade9f57686f31473e25be46ab282d9603f551401250f9f44
size: 32521
subdir: noarch
timestamp: 1668051714265
version: 0.38.4
fn: wheel-0.38.4-pyhd8ed1ab_0.tar.bz2
url: "https://conda.anaconda.org/conda-forge/noarch/wheel-0.38.4-pyhd8ed1ab_0.tar.bz2"
channel: "https://conda.anaconda.org/conda-forge/noarch"
package_tarball_full_path: "C:\\Users\\bas\\micromamba\\envs\\conda\\pkgs\\wheel-0.38.4-pyhd8ed1ab_0.tar.bz2"
extracted_package_dir: "C:\\Users\\bas\\micromamba\\envs\\conda\\pkgs\\wheel-0.38.4-pyhd8ed1ab_0"
files:
  - Lib/site-packages/wheel-0.38.4-py3.11.egg-info/PKG-INFO
  - Lib/site-packages/wheel-0.38.4-py3.11.egg-info/SOURCES.txt
  - Lib/site-packages/wheel-0.38.4-py3.11.egg-info/dependency_links.txt
  - Lib/site-packages/wheel-0.38.4-py3.11.egg-info/entry_points.txt
  - Lib/site-packages/wheel-0.38.4-py3.11.egg-info/not-zip-safe
  - Lib/site-packages/wheel-0.38.4-py3.11.egg-info/requires.txt
  - Lib/site-packages/wheel-0.38.4-py3.11.egg-info/top_level.txt
  - Lib/site-packages/wheel/__init__.py
  - Lib/site-packages/wheel/__main__.py
  - Lib/site-packages/wheel/_setuptools_logging.py
  - Lib/site-packages/wheel/bdist_wheel.py
  - Lib/site-packages/wheel/cli/__init__.py
  - Lib/site-packages/wheel/cli/convert.py
  - Lib/site-packages/wheel/cli/pack.py
  - Lib/site-packages/wheel/cli/unpack.py
  - Lib/site-packages/wheel/macosx_libfile.py
  - Lib/site-packages/wheel/metadata.py
  - Lib/site-packages/wheel/util.py
  - Lib/site-packages/wheel/vendored/__init__.py
  - Lib/site-packages/wheel/vendored/packaging/__init__.py
  - Lib/site-packages/wheel/vendored/packaging/_manylinux.py
  - Lib/site-packages/wheel/vendored/packaging/_musllinux.py
  - Lib/site-packages/wheel/vendored/packaging/tags.py
  - Lib/site-packages/wheel/wheelfile.py
  - Lib/site-packages/wheel/__pycache__/__init__.cpython-311.pyc
  - Lib/site-packages/wheel/__pycache__/__main__.cpython-311.pyc
  - Lib/site-packages/wheel/__pycache__/_setuptools_logging.cpython-311.pyc
  - Lib/site-packages/wheel/__pycache__/bdist_wheel.cpython-311.pyc
  - Lib/site-packages/wheel/cli/__pycache__/__init__.cpython-311.pyc
  - Lib/site-packages/wheel/cli/__pycache__/convert.cpython-311.pyc
  - Lib/site-packages/wheel/cli/__pycache__/pack.cpython-311.pyc
  - Lib/site-packages/wheel/cli/__pycache__/unpack.cpython-311.pyc
  - Lib/site-packages/wheel/__pycache__/macosx_libfile.cpython-311.pyc
  - Lib/site-packages/wheel/__pycache__/metadata.cpython-311.pyc
  - Lib/site-packages/wheel/__pycache__/util.cpython-311.pyc
  - Lib/site-packages/wheel/vendored/__pycache__/__init__.cpython-311.pyc
  - Lib/site-packages/wheel/vendored/packaging/__pycache__/__init__.cpython-311.pyc
  - Lib/site-packages/wheel/vendored/packaging/__pycache__/_manylinux.cpython-311.pyc
  - Lib/site-packages/wheel/vendored/packaging/__pycache__/_musllinux.cpython-311.pyc
  - Lib/site-packages/wheel/vendored/packaging/__pycache__/tags.cpython-311.pyc
  - Lib/site-packages/wheel/__pycache__/wheelfile.cpython-311.pyc
  - Scripts/wheel-script.py
  - Scripts/wheel.exe
paths_data:
  paths_version: 1
  paths:
    - _path: site-packages/wheel-0.38.4-py3.11.egg-info/PKG-INFO
      path_type: hardlink
      sha256: 552ac9fb5272cf729a7fcd2c6f77b546d93a8f69206ad247611bfd2485978557
      sha256_in_prefix: 552ac9fb5272cf729a7fcd2c6f77b546d93a8f69206ad247611bfd2485978557
      size_in_bytes: 2060
    - _path: site-packages/wheel-0.38.4-py3.11.egg-info/SOURCES.txt
      path_type: hardlink
      sha256: 5082fd55d385b979d4d1c28683b4d0b0f76c32f2d18dd0f260fb5e0a348bb589
      sha256_in_prefix: 5082fd55d385b979d4d1c28683b4d0b0f76c32f2d18dd0f260fb5e0a348bb589
      size_in_bytes: 2373
    - _path: site-packages/wheel-0.38.4-py3.11.egg-info/dependency_links.txt
      path_type: hardlink
      sha256: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b
      sha256_in_prefix: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b
      size_in_bytes: 1
    - _path: site-packages/wheel-0.38.4-py3.11.egg-info/entry_points.txt
      path_type: hardlink
      sha256: 92b83e88729e7e7b24d6abcd2c39193f7fb802adc9d05ff325a6ad861b742412
      sha256_in_prefix: 92b83e88729e7e7b24d6abcd2c39193f7fb802adc9d05ff325a6ad861b742412
      size_in_bytes: 107
    - _path: site-packages/wheel-0.38.4-py3.11.egg-info/not-zip-safe
      path_type: hardlink
      sha256: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b
      sha256_in_prefix: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b
      size_in_bytes: 1
    - _path: site-packages/wheel-0.38.4-py3.11.egg-info/requires.txt
      path_type: hardlink
      sha256: 56693d2a4c6ad40f15f029c092c0fd76abbfe8a08d45948f2b199e6dce39ce2b
      sha256_in_prefix: 56693d2a4c6ad40f15f029c092c0fd76abbfe8a08d45948f2b199e6dce39ce2b
      size_in_bytes: 22
    - _path: site-packages/wheel-0.38.4-py3.11.egg-info/top_level.txt
      path_type: hardlink
      sha256: 1f148121b804b2d30f7b87856b0840eba32af90607328a5756802771f8dbff57
      sha256_in_prefix: 1f148121b804b2d30f7b87856b0840eba32af90607328a5756802771f8dbff57
      size_in_bytes: 6
    - _path: site-packages/wheel/__init__.py
      path_type: hardlink
      sha256: db026b83eb702551c86d55ef7998f1c8cb718de9593957dff5b9e14c1b7779e7
      sha256_in_prefix: db026b83eb702551c86d55ef7998f1c8cb718de9593957dff5b9e14c1b7779e7
      size_in_bytes: 59
    - _path: site-packages/wheel/__main__.py
      path_type: hardlink
      sha256: 3643149ee4c219c3a4818d0804b8010950bf04619c58e471d8af236064b5d941
      sha256_in_prefix: 3643149ee4c219c3a4818d0804b8010950bf04619c58e471d8af236064b5d941
      size_in_bytes: 455
    - _path: site-packages/wheel/_setuptools_logging.py
      path_type: hardlink
      sha256: 3680a78c9e03144678e44a3ed817572ec5890b01a46a2b75b69ff5ee96a5795c
      sha256_in_prefix: 3680a78c9e03144678e44a3ed817572ec5890b01a46a2b75b69ff5ee96a5795c
      size_in_bytes: 746
    - _path: site-packages/wheel/bdist_wheel.py
      path_type: hardlink
      sha256: 93f81e7b6c98e137c3aff7513837320aca3a22da627532e3e8965f6ef67ddea9
      sha256_in_prefix: 93f81e7b6c98e137c3aff7513837320aca3a22da627532e3e8965f6ef67ddea9
      size_in_bytes: 19293
    - _path: site-packages/wheel/cli/__init__.py
      path_type: hardlink
      sha256: 35587cc7bf501b26f06692f954c42042ffb3c1283feae2ac6155db215464a50e
      sha256_in_prefix: 35587cc7bf501b26f06692f954c42042ffb3c1283feae2ac6155db215464a50e
      size_in_bytes: 2384
    - _path: site-packages/wheel/cli/convert.py
      path_type: hardlink
      sha256: b2451fe13b99724b2a1bbe49fbf2949055dd9980f14c9a4fdf5d4ed7a70d279d
      sha256_in_prefix: b2451fe13b99724b2a1bbe49fbf2949055dd9980f14c9a4fdf5d4ed7a70d279d
      size_in_bytes: 9427
    - _path: site-packages/wheel/cli/pack.py
      path_type: hardlink
      sha256: cd0d73989a2e37c63cea1089de54da2a1ea0ed26ea800d8cb8a3f5c18faf8dfc
      sha256_in_prefix: cd0d73989a2e37c63cea1089de54da2a1ea0ed26ea800d8cb8a3f5c18faf8dfc
      size_in_bytes: 3383
    - _path: site-packages/wheel/cli/unpack.py
      path_type: hardlink
      sha256: 414fce54c0af616b6b8d06b5f19e5929a67018122602a248cd0a2462ddaeddb2
      sha256_in_prefix: 414fce54c0af616b6b8d06b5f19e5929a67018122602a248cd0a2462ddaeddb2
      size_in_bytes: 659
    - _path: site-packages/wheel/macosx_libfile.py
      path_type: hardlink
      sha256: 39733a393c753bf0022c6044d90f69ac81e3e3bb969194994ae33fef9e96f3d4
      sha256_in_prefix: 39733a393c753bf0022c6044d90f69ac81e3e3bb969194994ae33fef9e96f3d4
      size_in_bytes: 16145
    - _path: site-packages/wheel/metadata.py
      path_type: hardlink
      sha256: fba9f5fa11fc62d99457ccecad17f6d3a8973c9256c84f6daa5a99bf95c6a528
      sha256_in_prefix: fba9f5fa11fc62d99457ccecad17f6d3a8973c9256c84f6daa5a99bf95c6a528
      size_in_bytes: 3727
    - _path: site-packages/wheel/util.py
      path_type: hardlink
      sha256: 7b48e99ec6db33d42169a312c9aa7efd9814c5cc70a722c393a44772b76e3cb8
      sha256_in_prefix: 7b48e99ec6db33d42169a312c9aa7efd9814c5cc70a722c393a44772b76e3cb8
      size_in_bytes: 621
    - _path: site-packages/wheel/vendored/__init__.py
      path_type: hardlink
      sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
      sha256_in_prefix: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
      size_in_bytes: 0
    - _path: site-packages/wheel/vendored/packaging/__init__.py
      path_type: hardlink
      sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
      sha256_in_prefix: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
      size_in_bytes: 0
    - _path: site-packages/wheel/vendored/packaging/_manylinux.py
      path_type: hardlink
      sha256: d4e58a003eb1b604e038654486e8c55aa177e495aa32f40bc59119d9094723d8
      sha256_in_prefix: d4e58a003eb1b604e038654486e8c55aa177e495aa32f40bc59119d9094723d8
      size_in_bytes: 11489
    - _path: site-packages/wheel/vendored/packaging/_musllinux.py
      path_type: hardlink
      sha256: 93dbd98f8b66c7412517edb2f21ee06f3d3dcc27e112367d6506de66be6c55db
      sha256_in_prefix: 93dbd98f8b66c7412517edb2f21ee06f3d3dcc27e112367d6506de66be6c55db
      size_in_bytes: 4374
    - _path: site-packages/wheel/vendored/packaging/tags.py
      path_type: hardlink
      sha256: 33f0d0238cc69cfab786c455f505987f94a830998107dd60aaa1bd26dcaffa6f
      sha256_in_prefix: 33f0d0238cc69cfab786c455f505987f94a830998107dd60aaa1bd26dcaffa6f
      size_in_bytes: 15612
    - _path: site-packages/wheel/wheelfile.py
      path_type: hardlink
      sha256: f6259639672f5574b1dba61d7caf50a6d440d4e1de4518ea128ecf11beb7d662
      sha256_in_prefix: f6259639672f5574b1dba61d7caf50a6d440d4e1de4518ea128ecf11beb7d662
      size_in_bytes: 7536
    - _path: Lib/site-packages/wheel/__pycache__/__init__.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/__pycache__/__main__.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/__pycache__/_setuptools_logging.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/__pycache__/bdist_wheel.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/cli/__pycache__/__init__.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/cli/__pycache__/convert.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/cli/__pycache__/pack.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/cli/__pycache__/unpack.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/__pycache__/macosx_libfile.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/__pycache__/metadata.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/__pycache__/util.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/vendored/__pycache__/__init__.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/vendored/packaging/__pycache__/__init__.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/vendored/packaging/__pycache__/_manylinux.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/vendored/packaging/__pycache__/_musllinux.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/vendored/packaging/__pycache__/tags.cpython-311.pyc
      path_type: pyc_file
    - _path: Lib/site-packages/wheel/__pycache__/wheelfile.cpython-311.pyc
      path_type: pyc_file
    - _path: Scripts/wheel-script.py
      path_type: windows_python_entry_point_script
    - _path: Scripts/wheel.exe
      path_type: windows_python_entry_point_exe
link:
  source: "C:\\Users\\bas\\micromamba\\envs\\conda\\pkgs\\wheel-0.38.4-pyhd8ed1ab_0"
  type: 1
requested_spec: "conda-forge/noarch::wheel==0.38.4=pyhd8ed1ab_0[md5=c829cfb8cb826acb9de0ac1a2df0a940]"
